File: src/utils/filters.py
from datetime import datetime, timedelta


def filter_by_date(df, preset):
    """Filter DataFrame based on date preset: All, Today, This Week, This Month."""
    if df.empty or preset == "All":
        return df

    today = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)

    if preset == "Today":
        return df[df["date_start"].dt.date == today.date()]
    elif preset == "This Week":
        return df[(df["date_start"] >= today) & (df["date_start"] < today + timedelta(days=7))]
    elif preset == "This Month":
        next_month = (today.replace(day=28) + timedelta(days=4)).replace(day=1)
        return df[(df["date_start"] >= today) & (df["date_start"] < next_month)]

    return df


def apply_filters(df, filters_state):
    """Apply all filters from state to DataFrame. Operates on views, avoids copies."""
    if df.empty:
        return df

    filtered = df

    # City
    if filters_state.city != "All Cities":
        filtered = filtered[filtered["city"] == filters_state.city]

    # Categories
    if filters_state.categories:
        filtered = filtered[filtered["category"].isin(filters_state.categories)]

    # Date
    filtered = filter_by_date(filtered, filters_state.date_preset)

    # Price
    if "price" in filtered.columns:
        filtered = filtered[filtered["price"] <= filters_state.max_price]

    # Search — combine into single mask operation
    if filters_state.search_query:
        q = filters_state.search_query.lower()
        mask = filtered["title"].str.lower().str.contains(q, na=False)
        if "venue" in filtered.columns:
            mask = mask | filtered["venue"].str.lower().str.contains(q, na=False)
        filtered = filtered[mask]

    return filtered

File: src/lib/filters.ts
import { Event, FilterState } from "@/types/event";
import { startOfDay, addDays, startOfMonth, addMonths } from "date-fns";

export function filterByDate(events: Event[], preset: string): Event[] {
  if (preset === "All") return events;

  const today = startOfDay(new Date());

  switch (preset) {
    case "Today":
      return events.filter(e => {
        const d = new Date(e.date_start);
        return d >= today && d < addDays(today, 1);
      });
    case "This Week":
      return events.filter(e => {
        const d = new Date(e.date_start);
        return d >= today && d < addDays(today, 7);
      });
    case "This Month": {
      const nextMonth = startOfMonth(addMonths(today, 1));
      return events.filter(e => {
        const d = new Date(e.date_start);
        return d >= today && d < nextMonth;
      });
    }
    default:
      return events;
  }
}

export function applyFilters(events: Event[], filters: FilterState): Event[] {
  let filtered = events;

  // City
  if (filters.city && filters.city !== "All Cities") {
    filtered = filtered.filter(e => e.city === filters.city);
  }

  // Categories
  if (filters.categories.length > 0) {
    filtered = filtered.filter(e => filters.categories.includes(e.category));
  }

  // Date
  filtered = filterByDate(filtered, filters.datePreset);

  // Price
  if (filters.maxPrice < 200) {
    filtered = filtered.filter(e => e.price <= filters.maxPrice);
  }

  // Search
  if (filters.searchQuery) {
    const q = filters.searchQuery.toLowerCase();
    filtered = filtered.filter(
      e =>
        e.title.toLowerCase().includes(q) ||
        e.venue.toLowerCase().includes(q)
    );
  }

  return filtered;
}

export function sortEvents(events: Event[], sortBy: string): Event[] {
  const sorted = [...events];
  switch (sortBy) {
    case "Date (Soonest)":
      return sorted.sort((a, b) => new Date(a.date_start).getTime() - new Date(b.date_start).getTime());
    case "Price (Low to High)":
      return sorted.sort((a, b) => a.price - b.price);
    case "Price (High to Low)":
      return sorted.sort((a, b) => b.price - a.price);
    case "Popularity":
      return sorted.sort((a, b) => b.popularity - a.popularity);
    default:
      return sorted;
  }
}
